core/src/KokkosExp_InterOp.hpp

```diff
@@ -0,0 +1,172 @@
+using concat_type_list_t = typename concat_type_list<T...>::type;
+
+// ------------------------------------------------------------------ //
+//  gather filters out types which satisfy PredicateT<T>::Value = ValueT
+
+template <template <typename> class PredicateT, bool ValueT, typename... T>
+struct gather {
+  using type = concat_type_list_t<std::conditional_t<
+      PredicateT<T>::value == ValueT, concat_type_list_t<T>, type_list<>>...>;
+};
+
+template <template <typename> class PredicateT, bool ValueT, typename... T>
+using gather_t = typename gather<PredicateT, ValueT, T...>::type;
+
+// ------------------------------------------------------------------ //
+//  this is used to convert
+//      Kokkos::Device<ExecSpace, MemSpace> to MemSpace
+//
+template <typename Tp>
+struct device_memory_space {
+  using type = Tp;
+};
+
+template <typename ExecT, typename MemT>
+struct device_memory_space<Kokkos::Device<ExecT, MemT>> {
+  using type = MemT;
+};
+
+template <typename Tp>
+using device_memory_space_t = typename device_memory_space<Tp>::type;
+
+// ------------------------------------------------------------------ //
+//  this identifies the default memory trait
+//
+template <typename Tp>
+struct is_default_memory_trait : std::false_type {};
+
+template <>
+struct is_default_memory_trait<Kokkos::MemoryTraits<0>> : std::true_type {};
+
+// ------------------------------------------------------------------ //
+//  this is the impl version which takes a view and converts to python
+//  view type
+//
+template <typename, typename...>
+struct python_view_type;
+
+template <template <typename...> class ViewT, typename ValueT,
+          typename... Types>
+struct python_view_type<ViewT<ValueT>, type_list<Types...>> {
+  using type = ViewT<ValueT, device_memory_space_t<Types>...>;
+};
+
+template <template <typename...> class ViewT, typename ValueT,
+          typename... Types>
+struct python_view_type<ViewT<ValueT, Types...>>
+    : python_view_type<ViewT<ValueT>,
+                       gather_t<is_default_memory_trait, false, Types...>> {};
+
+template <typename... T>
+using python_view_type_t = typename python_view_type<T...>::type;
+
+}  // namespace Impl
+
+//--------------------------------------------------------------------------------------//
+//  this is used to extract the uniform type of a view
+//
+template <typename ViewT>
+struct python_view_type {
+  static_assert(Kokkos::is_view<std::decay_t<ViewT>>::value ||
+                    Kokkos::is_dyn_rank_view<std::decay_t<ViewT>>::value,
+                "Error! python_view_type only supports Kokkos::View and "
+                "Kokkos::DynRankView");
+
+  using type = Impl::python_view_type_t<typename ViewT::array_type>;
+};
+
+template <typename ViewT>
+using python_view_type_t = typename python_view_type<ViewT>::type;
+
+template <typename Tp>
+auto as_python_type(Tp&& _v) {
+  using cast_type = python_view_type_t<Tp>;
+  return static_cast<cast_type>(std::forward<Tp>(_v));
+}
```

Comment:
Are you sure about decaying the view type?  Why is it OK to strip the reference?  What about the CV qualifiers?

Reply:
Where are the references and CV being stripped? I'm only using decay when testing the static assertion

Reply:
You're right.